handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java

```diff
@@ -608,70 +650,73 @@ public synchronized SSLEngineResult unwrap(
         }
 
         // Write encrypted data to network BIO
-        int bytesConsumed = -1;
-        try {
-            while (srcsOffset < srcsEndOffset) {
+        int bytesConsumed = 0;
+        if (srcsOffset < srcsEndOffset) {
+            do {
                 ByteBuffer src = srcs[srcsOffset];
                 int remaining = src.remaining();
                 int written = writeEncryptedData(src);
-                if (written >= 0) {
-                    if (bytesConsumed == -1) {
-                        bytesConsumed = written;
-                    } else {
-                        bytesConsumed += written;
-                    }
+                if (written > 0) {
+                    bytesConsumed += written;
+
                     if (written == remaining) {
                         srcsOffset ++;
-                    } else if (written == 0) {
-                        break;
                     }
                 } else {
+                    // BIO_write returned a negative or zero number, this means we could not complete the write
+                    // operation and should retry later.
+                    // TODO: Should we use BIO_should_retry() to get the exact cause ?
+                    SSL.clearError();
                     break;
                 }
-            }
-        } catch (Exception e) {
-            throw new SSLException(e);
-        }
-        if (bytesConsumed >= 0) {
+            } while (srcsOffset < srcsEndOffset);
+
             int lastPrimingReadResult = SSL.readFromSSL(ssl, EMPTY_ADDR, 0); // priming read
 
             // check if SSL_read returned <= 0. In this case we need to check the error and see if it was something
             // fatal.
             if (lastPrimingReadResult <= 0) {
-                // Check for OpenSSL errors caused by the priming read
-                long error = SSL.getLastErrorNumber();
-                if (OpenSsl.isError(error)) {
-                    String err = SSL.getErrorString(error);
-                    if (logger.isDebugEnabled()) {
-                        logger.debug(
-                                "SSL_read failed: primingReadResult: " + lastPrimingReadResult +
-                                        "; OpenSSL error: '" + err + '\'');
+                int sslError = SSL.getError(ssl, lastPrimingReadResult);
+                switch (sslError) {
+                case SSL.SSL_ERROR_NONE:
+                case SSL.SSL_ERROR_WANT_ACCEPT:
+                case SSL.SSL_ERROR_WANT_CONNECT:
+                case SSL.SSL_ERROR_WANT_WRITE:
+                case SSL.SSL_ERROR_WANT_READ:
+                case SSL.SSL_ERROR_WANT_X509_LOOKUP:
+                case SSL.SSL_ERROR_ZERO_RETURN:
```

Comment:
It's not really similar as how we handle the error is a bit different, so let us keep it like it is.